web/shared/src/typing_status.ts

```diff
@@ -1,5 +1,17 @@
 import _ from "lodash";
 
+type TypingStatusWorker = {
+    get_current_time: () => number;
+    notify_server_start: (recipient: number[] | null) => void;
+    notify_server_stop: (recipient: number[] | null) => void;
+};
+
+type TypingStatusState = {
+    current_recipient: number[] | null;
+    next_send_start_time: number | undefined;
+    idle_timer: ReturnType<typeof setTimeout> | undefined;
+};
```

Comment:
Are the situations where these things are not null/undefined coupled? I'm wondering if we should instead have the a single state object which is either null or has all these values set.

Reply:
Well, we progressively set fields of this `state` object and never set all the fields at once. So if we want to have a single state object then we would have incomplete types all over the place and will need some code refactoring to fix those errors?

Reply:
Given that we fall back to `0` when `next_send_start_time` is undefined, I think we don't even need to allow it to be `undefined`; we always call `start_or_extend_idle_timer` after setting `state.current_recipient`, where the former sets `idle_timer`.
It is feasible to refactor in a separate pre-ts commit to make it a lot cleaner.